diff -Naur PythonUpdate.orig/PythonUpdate-0.ebuild PythonUpdate/PythonUpdate-0.ebuild
--- PythonUpdate.orig/PythonUpdate-0.ebuild	2024-06-20 20:20:43.336566912 -0600
+++ PythonUpdate/PythonUpdate-0.ebuild	2024-06-20 20:47:16.957866040 -0600
@@ -4,7 +4,7 @@
 EAPI=8
 
 # unordered impls
-PYTHON_COMPAT=( python3_{9,8,10})
+PYTHON_COMPAT=( python3_{9,8,10,11})
 inherit python-r1
 
 DESCRIPTION="Ebuild with potential python updates"
diff -Naur PythonUpdate.orig/PythonUpdate-1.ebuild PythonUpdate/PythonUpdate-1.ebuild
--- PythonUpdate.orig/PythonUpdate-1.ebuild	2024-06-20 20:31:12.138216375 -0600
+++ PythonUpdate/PythonUpdate-1.ebuild	2024-06-20 20:47:27.817916266 -0600
@@ -3,7 +3,7 @@
 
 EAPI=8
 
-PYTHON_COMPAT=( python3_{8,9} )
+PYTHON_COMPAT=( python3_{8,9,10,11} )
 inherit python-single-r1
 
 DESCRIPTION="Ebuild with potential python updates"
diff -Naur PythonUpdate.orig/PythonUpdate-2.ebuild PythonUpdate/PythonUpdate-2.ebuild
--- PythonUpdate.orig/PythonUpdate-2.ebuild	2024-06-20 20:30:07.572879407 -0600
+++ PythonUpdate/PythonUpdate-2.ebuild	2024-06-20 20:47:36.831957939 -0600
@@ -3,7 +3,7 @@
 
 EAPI=8
 
-PYTHON_COMPAT=( python3_8 )
+PYTHON_COMPAT=( python3_{8,9,10,11} )
 inherit python-any-r1
 
 DESCRIPTION="Ebuild with potential python updates"
