diff -Naur WhitespaceUnneeded.orig/WhitespaceUnneeded-0.ebuild WhitespaceUnneeded/WhitespaceUnneeded-0.ebuild
--- WhitespaceUnneeded.orig/WhitespaceUnneeded-0.ebuild	2024-06-22 09:22:50.367865238 -0600
+++ WhitespaceUnneeded/WhitespaceUnneeded-0.ebuild	2024-06-22 09:26:09.956725639 -0600
@@ -1,10 +1,8 @@
 EAPI=8
 
-
 DESCRIPTION="Package with various unneeded whitespace usage"
-HOMEPAGE="https://pkgcraft.pkgcraft" 
-LICENSE="MIT"	
+HOMEPAGE="https://pkgcraft.pkgcraft"
+LICENSE="MIT"
 
 SLOT="0"
-	
 KEYWORDS="amd64"
