Index: pyproject.toml
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.BaseRevisionTextPatchEP
<+>[build-system]\nrequires = [\"hatchling\"]\nbuild-backend = \"hatchling.build\"\n\n[project]\nname = \"Pypable\"\nversion = \"1.0.0\"\nauthors = [\n  { name=\"Faelin\", email=\"faelin.codes@gmail.com\" },\n]\ndescription = \"A toolset for piping values between functions, with optional shell-like text-processing tools.\"\nreadme = \"README.md\"\nrequires-python = \">=3.7\"\nclassifiers = [\n    \"Programming Language :: Python :: 3\",\n    \"License :: OSI Approved :: MIT License\",\n    \"Operating System :: OS Independent\",\n]\n\n[project.urls]\nHomepage = \"https://github.com/faelin/Pypable\"\nIssues = \"https://github.com/faelin/Pypable/issues\"
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/pyproject.toml b/pyproject.toml
--- a/pyproject.toml	(revision cd8f554ed7f7063e5c1bf79fc5496f22a97b316b)
+++ b/pyproject.toml	(date 1722205850413)
@@ -4,7 +4,7 @@
 
 [project]
 name = "Pypable"
-version = "1.0.0"
+version = "1.0.2"
 authors = [
   { name="Faelin", email="faelin.codes@gmail.com" },
 ]
